### ISSUE TITLE
Plugin makes request to a non-existing endpoint on connect.woocommerce.com (174)

### ISSUE THREAD

**Issue by jorgeatorres** [woocommerce/woocommerce-paypal-payments#175]
After onboarding, the extension seems to make a request to `/ppc/v1/notifications/webhooks` on connect.woocommerce.com, but this endpoint doesn't exist in the connect.woocommerce.com implementation.

Here's the stacktrace of this call:

<img width="1226" alt="Screen Shot 2021-06-08 at 10 07 07" src="https://user-images.githubusercontent.com/184724/121535347-b22e8480-c9c7-11eb-857d-27a7dbabea68.png">

Is this expected?

**Comment by AlexP11223** [comment]
Yes, seems like a bug. Happens only when using manual credentials input.

**Comment by namiokuzono** [comment]
TL;DR - This still happens in 4.1.2, and it's not limited to manual credentials as the 2021 comment says — here's an OAuth log showing it. Root cause: api.host is resolved once per request from the is-connected flag, so on the request that performs the connection it's still pointing at the Woo partner proxy. Webhook registration fires later in that same request and 404s. The local webhook record gets deleted regardless, because delete_option sits outside the try/catch. [#3670](https://github.com/woocommerce/woocommerce-paypal-payments/issues/3670) is the same bug from a merchant; [#4609](https://github.com/woocommerce/woocommerce-paypal-payments/pull/4609) is a different one.

---
Still reproduces in **4.1.2**, and I think the scoping in the comment above is what has kept this quiet — this is **not** limited to manual credentials input.

Hit this on a live store last week via the standard OAuth connect flow (no manual credentials anywhere in the session). Log excerpt, redacted:

```
06:19:00 INFO  - Attempting OAuth login to PayPal... CONTEXT: {"sandbox":false,...}
06:19:00 DEBUG - POST https://api-m.paypal.com/v1/oauth2/token/            → 200
06:19:00 INFO  - Merchant successfully connected to PayPal
06:19:00 INFO  - Flushing API caches...
06:19:00 DEBUG - POST https://api-m.paypal.com/v1/oauth2/token?grant_type=client_credentials → 200
06:19:01 DEBUG - GET  https://api-m.paypal.com/v1/customer/partners/.../<merchant-id> → 200
06:19:01 DEBUG - Webhook operation 'register' started.
06:19:01 DEBUG - GET  https://api.woocommerce.com/integrations/ppc/v1/notifications/webhooks → 404
06:19:01 ERROR - Failed to delete webhooks: [Error] Not Found
06:19:01 INFO  - Webhooks deleted.
06:19:01 DEBUG - POST https://api.woocommerce.com/integrations/ppc/v1/notifications/webhooks → 404
06:19:01 ERROR - Failed to subscribe webhooks: [Error] Not Found
```

From the *next* request onward the same endpoint correctly targets `api-m.paypal.com` and returns 200:

```
07:29:54 DEBUG - [New Request] GET /wp-json/wc/v3/wc_paypal/webhooks
07:29:54 DEBUG - GET https://api-m.paypal.com/v1/notifications/webhooks → 200
```

### Root cause

`api.host` is a container service resolved **once per request** from the connection flag:

```php
// modules/ppcp-api-client/services.php:1019-1027
'api.production-host' => static function ( ContainerInterface $container ): string {
    $is_connected = $container->get( 'settings.flag.is-connected' );
    if ( $is_connected ) {
        return PAYPAL_API_URL;    // api-m.paypal.com
    }
    return CONNECT_WOO_URL;       // api.woocommerce.com/integrations/ppc
},
```

On the request that *performs* the connection, `is-connected` is still false when `api.host` first resolves, so it freezes to `CONNECT_WOO_URL` for the remainder of that request. `api.endpoint.webhook` consumes that cached string (`services.php:215-224`), and webhook registration fires later in the same request — against the partner proxy, which has no such endpoint.

This is why some calls in that same PID succeed while the webhook calls 404: the bearer and `PartnersEndpoint` are built by `api.factory.paypal-bearer`, documented at `services.php:186-188` as creating "a fresh bearer on every call to reflect the current connection." Plain container services don't get that treatment.

`woocommerce_paypal_payments_flush_api_cache` (`ApiModule.php:118-139`) fires right before this and looks like it should cover it, but it flushes four token/credential `Cache` objects only — `api.host` is a resolved string, not a cache entry, so it survives the flush.

### Why this is worse than a failed API call

`do_unregister()` runs `delete_option( self::KEY )` **outside** the try/catch (`WebhookRegistrar.php:129`), so a registration that failed end-to-end still wipes the local webhook record — and logs `Webhooks deleted.` while doing it. The store is left with no local webhook registration and no user-visible error; the only surfaced symptom is the settings screen, which reads live from PayPal and therefore still looks correct.

### Suggested direction

Resolving the host at call time rather than at container-resolution time would fix it, as would rebuilding `api.endpoint.webhook` (or invalidating `api.host`) as part of the existing post-connection flush.

### Related

- #3670 is the same defect reported from the merchant side in Sept 2025 — same 404 on the same endpoint, closed as a support request without being linked here.
- Distinct from #4604 / #4609. That work scopes deregistration by host inside `WebhookRegistrar` and adds the SSR "Webhook delivery URL" row, which is a good fix for a different problem. It does not touch host selection, so after it ships registration will still 404 against `api.woocommerce.com` on the onboarding request — just less destructively.
---
Investigated and written with Claude Code.

**Comment by Biont** [comment]
Closed by #4669
